File: java/src/test/java/myairtable/tests/TestOrmCrudViaTable.java
package myairtable.tests;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import myairtable.Airtable;
import myairtable.AirtableException;
import myairtable.AirtableQuery;
import myairtable.OrmTable;
import myairtable.PrimaryFields;
import myairtable.PrimaryModel;
import myairtable.PrimaryMultipleSelectOption;
import myairtable.PrimarySingleSelectOption;
import org.junit.jupiter.api.Test;

/**
 * J4.7 — Typed ORM CRUD via table methods. Parity with Kotlin TestOrmCrudViaTable / Swift
 * TestOrmCrudViaTable / Rust test_orm_crud_via_table.
 */
class TestOrmCrudViaTable {
  private final Airtable airtable = TestSetup.makeAirtable();

  @Test
  void primaryKeyOnlyCrudRoundTrip() {
    String primaryKey = TestSetup.primaryKey("OrmTable", "PKOnly");
    PrimaryModel created =
        airtable.primary().create(PrimaryModel.builder().primaryKey(primaryKey).build());
    assertNotNull(created.getId());
    assertEquals(primaryKey, created.getPrimaryKey());

    String recordId = created.getId();
    try {
      PrimaryModel fetched = airtable.primary().get(recordId);
      assertEquals(recordId, fetched.getId());
      assertEquals(primaryKey, fetched.getPrimaryKey());
      assertTrue(fetched.dirtyFields().isEmpty(), "fresh decode has no dirty fields");

      fetched.setPrimaryKey(primaryKey + " Updated");
      assertFalse(fetched.dirtyFields().isEmpty());
      PrimaryModel updated = airtable.primary().update(fetched);
      assertEquals(primaryKey + " Updated", updated.getPrimaryKey());

      airtable.primary().delete(recordId);
      assertThrows(
          AirtableException.class,
          () -> airtable.primary().get(recordId),
          "record should be gone after delete");
    } catch (Throwable e) {
      try {
        airtable.primary().delete(recordId);
      } catch (RuntimeException ignored) {
        // best-effort cleanup
      }
      throw e;
    }
  }

  @Test
  void allSimplePropertiesRoundTrip() {
    String primaryKey = TestSetup.primaryKey("OrmTable", "AllProps");
    PrimaryModel model =
        PrimaryModel.builder()
            .primaryKey(primaryKey)
            .singleLineText("Hello World")
            .longText("Long text content")
            .email("test@example.com")
            .url("https://example.com")
            .phoneNumber("555-1234")
            .checkbox(true)
            .numberInt(42.0)
            .numberFloat(3.14)
            .currencyInt(10.0)
            .currencyFloat(9.99)
            .percentInt(0.5)
            .percentFloat(0.333)
            .duration(Duration.ofSeconds(3600))
            .build();
    PrimaryModel created = airtable.primary().create(model);
    String recordId = created.getId();
    try {
      assertEquals(primaryKey, created.getPrimaryKey());
      assertEquals("Hello World", created.getSingleLineText());
      assertEquals("test@example.com", created.getEmail());
      assertEquals(true, created.getCheckbox());
      assertEquals(42.0, created.getNumberInt());
      assertEquals(3.14, created.getNumberFloat());
      assertEquals(9.99, created.getCurrencyFloat());
      assertEquals("https://example.com", created.getUrl());
      assertEquals("555-1234", created.getPhoneNumber());
      assertEquals(
          Duration.ofSeconds(3600), created.getDuration(), "duration wire unit is seconds");
      airtable.primary().delete(recordId);
    } catch (Throwable e) {
      try {
        airtable.primary().delete(recordId);
      } catch (RuntimeException ignored) {
        // best-effort cleanup
      }
      throw e;
    }
  }

  @Test
  void selectEnumsAndRatingRoundTrip() {
    String primaryKey = TestSetup.primaryKey("OrmTable", "Selects");
    PrimaryModel model =
        PrimaryModel.builder()
            .primaryKey(primaryKey)
            .singleSelect(PrimarySingleSelectOption.CHOICE_1)
            .multipleSelect(
                List.of(PrimaryMultipleSelectOption.OPTION_1, PrimaryMultipleSelectOption.OPTION_2))
            .rating(3L)
            .build();
    PrimaryModel created = airtable.primary().create(model);
    String recordId = created.getId();
    try {
      assertEquals(PrimarySingleSelectOption.CHOICE_1, created.getSingleSelect());
      assertEquals(
          List.of(PrimaryMultipleSelectOption.OPTION_1, PrimaryMultipleSelectOption.OPTION_2),
          created.getMultipleSelect());
      assertEquals(3L, created.getRating());

      PrimaryModel fetched = airtable.primary().get(recordId);
      assertEquals(PrimarySingleSelectOption.CHOICE_1, fetched.getSingleSelect());

      fetched.setSingleSelect(PrimarySingleSelectOption.CHOICE_2);
      fetched.setMultipleSelect(List.of(PrimaryMultipleSelectOption.OPTION_3));
      fetched.setRating(5L);
      PrimaryModel updated = airtable.primary().update(fetched);
      assertEquals(PrimarySingleSelectOption.CHOICE_2, updated.getSingleSelect());
      assertEquals(List.of(PrimaryMultipleSelectOption.OPTION_3), updated.getMultipleSelect());
      assertEquals(5L, updated.getRating());

      airtable.primary().delete(recordId);
    } catch (Throwable e) {
      try {
        airtable.primary().delete(recordId);
      } catch (RuntimeException ignored) {
        // best-effort cleanup
      }
      throw e;
    }
  }

  @Test
  void getQueryReturnsFilteredRecordSet() {
    String suite = TestSetup.primaryKey("OrmTable", "List");
    List<PrimaryModel> models = new ArrayList<>();
    for (int i = 1; i <= 3; i++) {
      models.add(PrimaryModel.builder().primaryKey(suite + " " + i).build());
    }
    List<PrimaryModel> created = airtable.primary().create(models);
    List<String> createdIds = created.stream().map(PrimaryModel::getId).toList();
    try {
      assertEquals(3, created.size());
      List<PrimaryModel> results =
          airtable
              .primary()
              .get(new AirtableQuery().withFormula("FIND(\"" + suite + "\", {Primary Key})"));
      assertEquals(3, results.size());
      assertEquals(
          java.util.Set.copyOf(createdIds),
          java.util.Set.copyOf(results.stream().map(PrimaryModel::getId).toList()));
      airtable.primary().deleteAll(createdIds);
    } catch (Throwable e) {
      try {
        airtable.primary().deleteAll(createdIds);
      } catch (RuntimeException ignored) {
        // best-effort cleanup
      }
      throw e;
    }
  }

  @Test
  void fieldSelectionAndMaxRecordsApply() {
    String suite = TestSetup.primaryKey("OrmTable", "Select");
    List<PrimaryModel> models = new ArrayList<>();
    for (int i = 1; i <= 3; i++) {
      models.add(
          PrimaryModel.builder().primaryKey(suite + " " + i).singleLineText("text " + i).build());
    }
    List<PrimaryModel> created = airtable.primary().create(models);
    List<String> createdIds = created.stream().map(PrimaryModel::getId).toList();
    try {
      List<PrimaryModel> results =
          airtable
              .primary()
              .get(
                  new AirtableQuery()
                      .withFormula("FIND(\"" + suite + "\", {Primary Key})")
                      .withFields(List.of(PrimaryFields.primaryKeyId))
                      .withMaxRecords(2));
      assertEquals(2, results.size(), "maxRecords caps the result set");
      for (PrimaryModel model : results) {
        assertNotNull(model.getPrimaryKey());
        assertNull(model.getSingleLineText(), "unselected fields decode as null");
      }
      airtable.primary().deleteAll(createdIds);
    } catch (Throwable e) {
      try {
        airtable.primary().deleteAll(createdIds);
      } catch (RuntimeException ignored) {
        // best-effort cleanup
      }
      throw e;
    }
  }

  @Test
  void invalidRecordIdThrows() {
    assertThrows(AirtableException.class, () -> airtable.primary().get("recDOESNOTEXIST123"));
  }

  @Test
  void batchCreateUpdateDeleteChunksAboveTheBatchLimit() {
    String suite = TestSetup.primaryKey("OrmTable", "Batch");
    int count = 111;
    List<PrimaryModel> models = new ArrayList<>();
    for (int i = 1; i <= count; i++) {
      models.add(PrimaryModel.builder().primaryKey(suite + " " + i).build());
    }
    // create() chunks into 10-record batches; a mid-batch failure throws WITHOUT returning the
    // ids of records already created in earlier batches, so the by-id cleanup below can't see
    // them. Pull the create inside the try and fall back to a prefix sweep on failure (the suite
    // prefix is unique per run), so partial creates are still cleaned up.
    List<String> createdIds = List.of();
    try {
      List<PrimaryModel> created = airtable.primary().create(models);
      createdIds = created.stream().map(PrimaryModel::getId).toList();

      assertEquals(count, created.size());
      for (int i = 0; i < created.size(); i++) {
        assertEquals(suite + " " + (i + 1), created.get(i).getPrimaryKey());
      }

      for (PrimaryModel model : created) {
        model.setPrimaryKey(model.getPrimaryKey() + " Updated");
      }
      List<PrimaryModel> updated = airtable.primary().update(created);
      assertEquals(count, updated.size());
      for (int i = 0; i < updated.size(); i++) {
        assertEquals(suite + " " + (i + 1) + " Updated", updated.get(i).getPrimaryKey());
      }

      airtable.primary().deleteAll(createdIds);
      String firstId = createdIds.get(0);
      assertThrows(AirtableException.class, () -> airtable.primary().get(firstId));
    } catch (Throwable e) {
      // First delete the ids we know about, then sweep by prefix to catch any partial-batch
      // creates whose ids never came back (the original primary key carries the unique suite
      // prefix; the " Updated" rename only happens after a fully successful create).
      try {
        if (!createdIds.isEmpty()) {
          airtable.primary().deleteAll(createdIds);
        }
      } catch (RuntimeException ignored) {
        // best-effort cleanup
      }
      try {
        List<String> straySweep =
            airtable
                .primary()
                .get(new AirtableQuery().withFormula("FIND(\"" + suite + "\", {Primary Key})"))
                .stream()
                .map(PrimaryModel::getId)
                .toList();
        if (!straySweep.isEmpty()) {
          airtable.primary().deleteAll(straySweep);
        }
      } catch (RuntimeException ignored) {
        // best-effort cleanup
      }
      throw e;
    }
  }

  @Test
  void upsertInsertsWhenNoMatchThenUpdatesOnMatch() {
    String primaryKey = TestSetup.primaryKey("OrmTable", "Upsert");
    // Insert path
    OrmTable.UpsertResult<PrimaryModel> first =
        airtable
            .primary()
            .upsert(
                PrimaryModel.builder().primaryKey(primaryKey).build(),
                List.of(PrimaryFields.primaryKeyId));
    String recordId = first.model().getId();
    try {
      assertTrue(first.wasCreated(), "no match yet — upsert inserts");
      assertEquals(primaryKey, first.model().getPrimaryKey());

      // Update path: same primary key, new field value.
      OrmTable.UpsertResult<PrimaryModel> second =
          airtable
              .primary()
              .upsert(
                  PrimaryModel.builder().primaryKey(primaryKey).singleLineText("merged").build(),
                  List.of(PrimaryFields.primaryKeyId));
      assertFalse(second.wasCreated(), "matched on primary key — upsert updates");
      assertEquals(recordId, second.model().getId());
      assertEquals("merged", second.model().getSingleLineText());

      airtable.primary().delete(recordId);
    } catch (Throwable e) {
      try {
        airtable.primary().delete(recordId);
      } catch (RuntimeException ignored) {
        // best-effort cleanup
      }
      throw e;
    }
  }
}

File: kotlin/src/test/kotlin/myairtable/tests/TestOrmCrudViaTable.kt
package myairtable.tests

import kotlinx.coroutines.runBlocking
import kotlinx.serialization.json.JsonPrimitive
import myairtable.AirtableException
import myairtable.AirtableQuery
import myairtable.PrimaryFields
import myairtable.PrimaryModel
import myairtable.PrimaryMultipleSelectOption
import myairtable.PrimarySingleSelectOption
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertFalse
import kotlin.test.assertNotEquals
import kotlin.test.assertNotNull
import kotlin.test.assertTrue
import kotlin.time.Duration.Companion.seconds

/**
 * K4.7 — Typed ORM CRUD via table methods. Parity with Swift
 * TestOrmCrudViaTable / Rust test_orm_crud_via_table.
 */
class TestOrmCrudViaTable {
    private val airtable = TestSetup.makeAirtable()

    @Test
    fun primaryKeyOnlyCrudRoundTrip() =
        runBlocking {
            val primaryKey = TestSetup.primaryKey("OrmTable", "PKOnly")
            val created = airtable.primary.create(PrimaryModel(primaryKey = primaryKey))
            assertNotNull(created.id)
            assertEquals(primaryKey, created.primaryKey)

            val recordId = created.id!!
            try {
                val fetched = airtable.primary.get(recordId)
                assertEquals(recordId, fetched.id)
                assertEquals(primaryKey, fetched.primaryKey)
                assertTrue(fetched.dirtyFields().isEmpty(), "fresh decode has no dirty fields")

                fetched.primaryKey = "$primaryKey Updated"
                assertFalse(fetched.dirtyFields().isEmpty())
                val updated = airtable.primary.update(fetched)
                assertEquals("$primaryKey Updated", updated.primaryKey)

                airtable.primary.delete(recordId)
                val deleted = runCatching { airtable.primary.get(recordId) }.isFailure
                assertTrue(deleted)
            } catch (e: Throwable) {
                runCatching { airtable.primary.delete(recordId) }
                throw e
            }
        }

    @Test
    fun allSimplePropertiesRoundTrip() =
        runBlocking {
            val primaryKey = TestSetup.primaryKey("OrmTable", "AllProps")
            val model =
                PrimaryModel(
                    primaryKey = primaryKey,
                    singleLineText = "Hello World",
                    longText = "Long text content",
                    email = "test@example.com",
                    url = "https://example.com",
                    phoneNumber = "555-1234",
                    checkbox = true,
                    numberInt = 42.0,
                    numberFloat = 3.14,
                    currencyInt = 10.0,
                    currencyFloat = 9.99,
                    percentInt = 0.5,
                    percentFloat = 0.333,
                    duration = 3600.seconds,
                )
            val created = airtable.primary.create(model)
            val recordId = created.id!!
            try {
                assertEquals(primaryKey, created.primaryKey)
                assertEquals("Hello World", created.singleLineText)
                assertEquals("test@example.com", created.email)
                assertEquals(true, created.checkbox)
                assertEquals(42.0, created.numberInt)
                assertEquals(3.14, created.numberFloat)
                assertEquals(9.99, created.currencyFloat)
                assertEquals("https://example.com", created.url)
                assertEquals("555-1234", created.phoneNumber)
                assertEquals<kotlin.time.Duration?>(3600.seconds, created.duration, "duration wire unit is seconds")
                airtable.primary.delete(recordId)
            } catch (e: Throwable) {
                runCatching { airtable.primary.delete(recordId) }
                throw e
            }
        }

    @Test
    fun selectEnumsAndRatingRoundTrip() =
        runBlocking {
            val primaryKey = TestSetup.primaryKey("OrmTable", "Selects")
            val model =
                PrimaryModel(
                    primaryKey = primaryKey,
                    singleSelect = PrimarySingleSelectOption.CHOICE_1,
                    multipleSelect = listOf(PrimaryMultipleSelectOption.OPTION_1, PrimaryMultipleSelectOption.OPTION_2),
                    rating = 3L,
                )
            val created = airtable.primary.create(model)
            val recordId = created.id!!
            try {
                assertEquals(PrimarySingleSelectOption.CHOICE_1, created.singleSelect)
                assertEquals(listOf(PrimaryMultipleSelectOption.OPTION_1, PrimaryMultipleSelectOption.OPTION_2), created.multipleSelect)
                assertEquals(3L, created.rating)

                val fetched = airtable.primary.get(recordId)
                assertEquals(PrimarySingleSelectOption.CHOICE_1, fetched.singleSelect)

                fetched.singleSelect = PrimarySingleSelectOption.CHOICE_2
                fetched.multipleSelect = listOf(PrimaryMultipleSelectOption.OPTION_3)
                fetched.rating = 5L
                val updated = airtable.primary.update(fetched)
                assertEquals(PrimarySingleSelectOption.CHOICE_2, updated.singleSelect)
                assertEquals(listOf(PrimaryMultipleSelectOption.OPTION_3), updated.multipleSelect)
                assertEquals(5L, updated.rating)

                airtable.primary.delete(recordId)
            } catch (e: Throwable) {
                runCatching { airtable.primary.delete(recordId) }
                throw e
            }
        }

    @Test
    fun getQueryReturnsFilteredRecordSet() =
        runBlocking {
            val suite = TestSetup.primaryKey("OrmTable", "List")
            val models = (1..3).map { PrimaryModel(primaryKey = "$suite $it") }
            val created = airtable.primary.create(models)
            val createdIds = created.mapNotNull { it.id }
            try {
                assertEquals(3, created.size)
                val results = airtable.primary.get(AirtableQuery(formula = """FIND("$suite", {Primary Key})"""))
                assertEquals(3, results.size)
                assertEquals(createdIds.toSet(), results.mapNotNull { it.id }.toSet())
                airtable.primary.delete(createdIds)
            } catch (e: Throwable) {
                runCatching { airtable.primary.delete(createdIds) }
                throw e
            }
        }

    @Test
    fun fieldSelectionAndMaxRecordsApply() =
        runBlocking {
            val suite = TestSetup.primaryKey("OrmTable", "Select")
            val models = (1..3).map { PrimaryModel(primaryKey = "$suite $it", singleLineText = "text $it") }
            val created = airtable.primary.create(models)
            val createdIds = created.mapNotNull { it.id }
            try {
                val results =
                    airtable.primary.get(
                        AirtableQuery(
                            formula = """FIND("$suite", {Primary Key})""",
                            fields = listOf(PrimaryFields.primaryKeyId),
                            maxRecords = 2,
                        ),
                    )
                assertEquals(2, results.size, "maxRecords caps the result set")
                for (model in results) {
                    assertNotNull(model.primaryKey)
                    kotlin.test.assertNull(model.singleLineText, "unselected fields decode as null")
                }
                airtable.primary.delete(createdIds)
            } catch (e: Throwable) {
                runCatching { airtable.primary.delete(createdIds) }
                throw e
            }
        }

    @Test
    fun invalidRecordIdThrows(): Unit =
        runBlocking {
            assertFailsWith<AirtableException> {
                airtable.primary.get("recDOESNOTEXIST123")
            }
        }

    @Test
    fun batchCreateUpdateDeleteChunksAboveTheBatchLimit() =
        runBlocking {
            val suite = TestSetup.primaryKey("OrmTable", "Batch")
            val count = 111
            val models = (1..count).map { PrimaryModel(primaryKey = "$suite $it") }
            val created = airtable.primary.create(models)
            val createdIds = created.mapNotNull { it.id }
            try {
                assertEquals(count, created.size)
                created.forEachIndexed { i, model -> assertEquals("$suite ${i + 1}", model.primaryKey) }

                for (model in created) model.primaryKey = "${model.primaryKey} Updated"
                val updated = airtable.primary.update(created)
                assertEquals(count, updated.size)
                updated.forEachIndexed { i, model -> assertEquals("$suite ${i + 1} Updated", model.primaryKey) }

                airtable.primary.delete(createdIds)
                assertFailsWith<AirtableException> { airtable.primary.get(createdIds.first()) }
            } catch (e: Throwable) {
                runCatching { airtable.primary.delete(createdIds) }
                throw e
            }
            Unit
        }

    @Test
    fun upsertInsertsWhenNoMatchThenUpdatesOnMatch() =
        runBlocking {
            val primaryKey = TestSetup.primaryKey("OrmTable", "Upsert")
            // Insert path
            val first = airtable.primary.upsert(PrimaryModel(primaryKey = primaryKey), listOf(PrimaryFields.primaryKeyId))
            val recordId = first.model.id!!
            try {
                assertTrue(first.wasCreated, "no match yet — upsert inserts")
                assertEquals(primaryKey, first.model.primaryKey)

                // Update path: same primary key, new field value.
                val second =
                    airtable.primary.upsert(
                        PrimaryModel(primaryKey = primaryKey, singleLineText = "merged"),
                        listOf(PrimaryFields.primaryKeyId),
                    )
                assertFalse(second.wasCreated, "matched on primary key — upsert updates")
                assertEquals(recordId, second.model.id)
                assertEquals("merged", second.model.singleLineText)

                airtable.primary.delete(recordId)
            } catch (e: Throwable) {
                runCatching { airtable.primary.delete(recordId) }
                throw e
            }
        }

    // Duplicate — parity with the other language duplicate suites.
    @Test
    fun duplicateCopiesWritableFieldsAndRecomputesTheRest() =
        runBlocking {
            val primaryKey = TestSetup.primaryKey("Duplicate", "Source")
            val source =
                airtable.primary.create(
                    PrimaryModel(
                        primaryKey = primaryKey,
                        singleLineText = "copy me",
                        // Stays <= 10 and != 20 on purpose: filter-by-formula asserts exact counts
                        // for `numberInt = 20` and `AND(numberInt > 10, checkbox = true)`.
                        numberInt = 7.0,
                        rating = 3L,
                    ),
                )
            val sourceId = source.id!!
            val copy = airtable.primary.duplicate(source)
            val copyId = copy.id!!
            try {
                assertNotEquals(sourceId, copyId)
                assertEquals(primaryKey, copy.primaryKey)
                assertEquals("copy me", copy.singleLineText)
                assertEquals(7.0, copy.numberInt)
                assertEquals(3L, copy.rating)
                // Formula (ID) resolves to RECORD_ID(), so on a true copy it is the COPY's id --
                // proof the computed fields were recalculated rather than copied.
                assertEquals(copyId, copy.formulaId?.valueOrNull)
                assertNotEquals(source.autoNumber?.valueOrNull, copy.autoNumber?.valueOrNull)
            } finally {
                airtable.primary.delete(listOf(sourceId, copyId))
            }
        }

    @Test
    fun duplicateByIdsPreservesInputOrder() =
        runBlocking {
            val aKey = TestSetup.primaryKey("Duplicate", "OrderA")
            val bKey = TestSetup.primaryKey("Duplicate", "OrderB")
            val a = airtable.primary.create(PrimaryModel(primaryKey = aKey))
            val b = airtable.primary.create(PrimaryModel(primaryKey = bKey))
            val copies = airtable.primary.duplicate(listOf(b.id!!, a.id!!))
            try {
                assertEquals(2, copies.size)
                assertEquals(bKey, copies[0].primaryKey)
                assertEquals(aKey, copies[1].primaryKey)
            } finally {
                airtable.primary.delete(listOf(a.id!!, b.id!!) + copies.map { it.id!! })
            }
        }
}
